### DISCUSSION TITLE
[Unstoppable] TotalSupply

### DISCUSSION THREAD

**Discussion by MartinTochon** [tinchoabbate/damn-vulnerable-defi#28] [Help / Q&A / Troubleshooting] [answered]
Hello everyone !

I'm kind of new to Web 3.0 and I recently discovered Damn VUlenrable DeFi.
It was only natural that I started with the "Unstoppable" challenge 1. However, I can't figure out where the `TotalSupply `variable used on line 96 of "UnstoppableVault.sol" is defined. This prevents me from understanding the solution to the problem because I can't see how `convertToShares(TotalSupply)` can be different from `balanceBefore` when to me the two represent the same thing.

Thanks for helping me :)

**Comment by Koowah** [comment]
Hi @MartinTochon ,

Are you using VSCode ? If so you can `ctrl (or cmd) + click` (or right click + go to definition) on the variable to go to its definition. This (most of the time) even works cross-file.

In your specific case, `totalSupply` is defined in the `ERC20` contract, which is inherited by `ERC4626`, which is itself inherited by the `UnstoppableVault`.

Hope that helps,
Cheers 😄

**Reply by MartinTochon** [reply]
Hey @Koowah,

Thanks a lot for your help. :) I understand better now! 
I still have a small question though, `convertToShares(TotalSupply) `are `shares` while `balanceBefore` represents `assets` as it is equal to `totalAssets()`. Can we really compare the two even if they do not represent the same thing?
I understand that both are uint256, but still, I don't see how it is relevant to compare the two.
Am I missing something again?

I hope I have made myself clear.
Thank you again for your help

**Reply by Koowah** [reply, marked as answer]
@MartinTochon 

That's a fair question. I didn't understand this either initially.

`ERC4626` is the standard for `tokenized vaults`. What's a tokenized vault ? Think about an ETH lending pool. People can deposit ETH (or rather WETH in our case) into the pool for other people to borrow (asset), and get an ERC20 that represents their shares (the proportion of the total lended eth in that pool) that can at any time be exchanged back for their original ETH. (the incentive is that they accrue yield based on their shares but this isn't relevant here)

The constructor of ERC4626 therefore takes 3 arguments : 
- An ERC20 : the underlying asset token (the thing that has value)
- A name and a symbol to create the shares token (the thing that can be exchanged for the asset and that is used internally by the protocol)

Consequently, `totalSupply`/`totalAssets` represent the amount shares/assets available, and both quantities are supposedly strongly linked, in the `ERC4626` standard, by the convertToShares and convertToAsset functions.

Just as a side note and advice for the following challenges : DamnVulnerableDefi is more demanding than Ethernaut in term of preparation before tackling each challenge. It requires a good level of familiarity with protocols rather than contract patterns; you will have to read through documentation, articles and watch videos if necessary to grasp the architecture of the projects which will take its fair share of time.

@tinchoabbate Just a thought but you could perhaps mention this last point in the `tips` section of the DamnVulnerableDefi homepage. I know it seems obvious but people (myself included) often expect the solutions in CTFs to be just under their nose when they might actually need to take a deeper look at all the available tooling.

**Reply by MartinTochon** [reply]
Thank you very much for your very detailed reply.
I'm not going to lie, I had to reread your message several times to understand it properly but thanks to you I understood the ERC4626 a little better :)

**Reply by Koowah** [reply]
@MartinTochon 
My pleasure. Feel free to ask if you need more explanations. It was tricky to assess how much detail to include in my previous answer 😄
